File: cloudera-framework-example/src/main/java/com/cloudera/framework/example/Driver.java
package com.cloudera.framework.example;

import java.util.Set;

import com.cloudera.framework.common.util.DfsUtil;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Driver super class
 */
public abstract class Driver extends com.cloudera.framework.common.Driver {

  private static final Logger LOG = LoggerFactory.getLogger(Driver.class);

  @SuppressWarnings("FieldCanBeLocal")
  protected Path inputPath;
  protected Path outputPath;
  protected Set<Path> inputPaths;

  @SuppressWarnings("FieldCanBeLocal")
  protected FileSystem hdfs;

  public Driver() {
    super();
  }

  public Driver(Configuration configuration) {
    super(configuration);
  }

  @Override
  public int prepare(String... arguments) throws Exception {
    if (arguments == null || arguments.length != 2) {
      throw new Exception("Invalid number of arguments");
    }
    hdfs = FileSystem.newInstance(getConf());
    inputPath = hdfs.makeQualified(new Path(arguments[0]));
    if (LOG.isInfoEnabled()) {
      LOG.info("Input path [" + inputPath + "] validated");
    }
    inputPaths = DfsUtil.listDirs(hdfs, inputPath, true, true);
    outputPath = hdfs.makeQualified(new Path(arguments[1]));
    hdfs.mkdirs(outputPath.getParent());
    if (LOG.isInfoEnabled()) {
      LOG.info("Output path [" + outputPath + "] validated");
    }
    return RETURN_SUCCESS;
  }

}

File: cloudera-framework-example/cloudera-framework-example-3/src/main/scala/com/cloudera/framework/example/three/Driver.scala
package com.cloudera.framework.example.three

import java.util.Properties

import com.cloudera.framework.common.Driver.RETURN_SUCCESS
import com.cloudera.framework.example.three.Driver.{ModelDir, TestDir, TrainDir}
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileSystem, Path}
import org.slf4j.LoggerFactory

import scala.io.Source

object Driver {

  val TestDir = "test"
  val TrainDir = "train"
  val ModelDir = "model"

  val ModelFile = "occupancy.pmml"
}

class Driver extends com.cloudera.framework.common.Driver {

  val properties = new Properties()
  private val Log = LoggerFactory.getLogger(classOf[Driver])
  var testPath: Path = _
  var trainPath: Path = _
  var modelPath: Path = _

  def this(configuration: Configuration) {
    this
    super.setConf(configuration)
  }

  def main(arguments: Array[String]): Unit = {
    System.exit(new Driver().runner(arguments))
  }

  override def prepare(arguments: String*): Int = {
    if (arguments == null || arguments.length != parameters().length) throw new Exception("Invalid number of arguments")
    properties.load(Source.fromURL(getClass.getResource("/application.properties")).bufferedReader())
    val hdfs = FileSystem.newInstance(getConf)
    val workingPath = hdfs.makeQualified(new Path(arguments(0)))
    if (!hdfs.exists(workingPath)) throw new Exception("Input path [" + workingPath + "] does not exist")
    if (Log.isInfoEnabled) Log.info("Working path [" + workingPath + "] validated")
    testPath = new Path(workingPath, TestDir)
    trainPath = new Path(workingPath, TrainDir)
    modelPath = new Path(workingPath, ModelDir)
    RETURN_SUCCESS
  }

  override def parameters(): Array[String] = {
    Array("working-path")
  }

  override def execute(): Int = {
    addResult(Model.build(FileSystem.newInstance(getConf), properties.getProperty("application.version"),
      trainPath.toString, testPath.toString, modelPath.toString).getOrElse(ModelPmml.EmptyModel))
    RETURN_SUCCESS
  }

}
